### PR TITLE
Integrate a local copy of rust-multiaddr for development

### PR DESCRIPTION
**Pull request by tomaka**
I've been working on improving `rust-multiaddr`, and have opened [an upstream PR](https://github.com/multiformats/rust-multiaddr/pull/28).
However:
- This PR is likely going to take a long time to get reviewed, as it contains big design changes.
- Some necessary changes and improvements may have been missed.

Therefore this PR integrates `rust-multiaddr` to the current repository, with the intent to remove it in the future after its design has been fleshed out.

Also modifies `tcp-transport` to use the new design.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,5 +14,5 @@ members = [
 [replace]
 # Ring has a feature merged in master that hasn't been published yet ; remove this after 0.12.2
 "ring:0.12.1" = { git = "https://github.com/briansmith/ring" }
-# Multiaddr has a feature merged in master that hasn't been published yet ; remove this after 0.2.1
-"multiaddr:0.2.0" = { git = "https://github.com/multiformats/rust-multiaddr" }
+# Using a local improved version of multiaddr for now
```
Comment (Vurich) [inline Cargo.toml]: _**I M P R O V E D**_

File: rust-multiaddr/src/errors.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+pub type Result<T> = ::std::result::Result<T, Error>;
+
+/// Error types
+#[derive(Debug)]
+pub enum Error {
+    UnknownProtocol,
+    UnknownProtocolString,
+    InvalidMultiaddr,
+    MissingAddress,
+    ParsingError(Box<error::Error + Send + Sync>),
```
Comment (Vurich) [inline rust-multiaddr/src/errors.rs]: Can we make this type more specific since there's a finite set of errors it could be? Possibly `error_chain` would make this nicer. If this isn't your code then we can do this as another PR.

File: rust-multiaddr/src/errors.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Error::UnknownProtocol => "unknown protocol",
+            Error::UnknownProtocolString => "unknown protocol string",
+            Error::InvalidMultiaddr => "invalid multiaddr",
+            Error::MissingAddress => "protocol requires address, none given",
+            Error::ParsingError(_) => "failed to parse",
```
Comment (Vurich) [inline rust-multiaddr/src/errors.rs]: Can we delegate to the inner error here?

File: rust-multiaddr/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+    fn to_string(&self) -> String {
+        let mut out = String::new();
+        for s in self.iter() {
+            out.push_str(&s.to_string());
+        }
+        out
+    }
+}
+
+impl fmt::Debug for Multiaddr {
```
Comment (Vurich) [inline rust-multiaddr/src/lib.rs]: Can we implement `Display` too and have `to_string` just be `format!("{}", self)`?

File: rust-multiaddr/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+    /// use multiaddr::Multiaddr;
+    ///
+    /// let address = Multiaddr::new("/ip4/127.0.0.1/udp/1234").unwrap();
+    /// assert_eq!(address.to_bytes(), [
+    ///     4, 127, 0, 0, 1,
+    ///     17, 4, 210
+    /// ]);
+    /// ```
+    ///
+    pub fn new(input: &str) -> Result<Multiaddr> {
```
Comment (Vurich) [inline rust-multiaddr/src/lib.rs]: I'd make this `fn parse`, to me a `new` function implies a minimal amount of work done.

File: rust-multiaddr/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+        let mut bytes = Vec::new();
+
+        let mut parts = input.split('/');
+        if !parts.next().ok_or(Error::InvalidMultiaddr)?.is_empty() {
```
Comment (Vurich) [inline rust-multiaddr/src/lib.rs]: Can we comment this with "Multiaddr must start with \`/\`" or something?

File: rust-multiaddr/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+                ProtocolArgSize::Fixed { bytes: 0 } => {
+                    protocol.parse_data("")?     // TODO: bad design
+                },
+                _ => {
+                    let data = parts.next().ok_or(Error::MissingAddress)?;
+                    protocol.parse_data(data)?
+                },
+            };
+
+            addr_component.write_bytes(&mut bytes).expect("writing to a Vec never fails");
```
Comment (Vurich) [inline rust-multiaddr/src/lib.rs]: If we know it's a `Vec` can we use `extend` instead?

File: rust-multiaddr/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+    ///
+    /// ```
+    /// use multiaddr::{Multiaddr, ProtocolId};
+    ///
+    /// let address = Multiaddr::new("/ip4/127.0.0.1").unwrap();
+    /// assert_eq!(address.protocol(), vec![ProtocolId::IP4]);
+    /// ```
+    ///
+    #[inline]
+    pub fn protocol(&self) -> Vec<ProtocolId> {
```
Comment (Vurich) [inline rust-multiaddr/src/lib.rs]: Can we make this `to_protocols` to make it obvious that it's doing work? Even better, make it `protocols` and return an `impl Iterator<Item = &ProtocolId>`

File: rust-multiaddr/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+    pub fn encapsulate<T: ToMultiaddr>(&self, input: T) -> Result<Multiaddr> {
+        let new = input.to_multiaddr()?;
+        let mut bytes = self.bytes.clone();
+
+        bytes.extend(new.to_bytes());
+
+        Ok(Multiaddr { bytes: bytes })
+    }
+
+    /// Remove the outer most address from itself.
```
Comment (Vurich) [inline rust-multiaddr/src/lib.rs]: `s/outer most/outermost/`. Also, "from itself" is unnecessary.

File: rust-multiaddr/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+    /// let address: Multiaddr = "/ip4/127.0.0.1/udt/sctp/5678".parse().unwrap();
+    /// 
+    /// let components = address.iter().collect::<Vec<_>>();
+    /// assert_eq!(components[0], AddrComponent::IP4(Ipv4Addr::new(127, 0, 0, 1)));
+    /// assert_eq!(components[1], AddrComponent::UDT);
+    /// assert_eq!(components[2], AddrComponent::SCTP(5678));
+    /// ```
+    ///
+    #[inline]
+    pub fn iter(&self) -> Iter {
```
Comment (Vurich) [inline rust-multiaddr/src/lib.rs]: If we have this method then we should implement `IntoIterator` for `&Multiaddr`

**Review by eira-fransham — CHANGES_REQUESTED** [review]
I don't know how much of this is your code and how much is existing, but I've commented on both.

Also, can we rename this to `multiaddr-rs`? That's the "Rust standard" naming convention.

File: rust-multiaddr/src/errors.rs (hunk 3 above)
Comment (tomaka) [reply rust-multiaddr/src/errors.rs]: That's not the idiomatic thing to do, although I agree that the idiomatic thing to do is not really great.

File: rust-multiaddr/src/lib.rs (hunk 5 above)
Comment (tomaka) [reply rust-multiaddr/src/lib.rs]: `Multiaddr` implements `FromStr`, so I'll just deprecate `new`.

File: rust-multiaddr/src/lib.rs (hunk 7 above)
Comment (tomaka) [reply rust-multiaddr/src/lib.rs]: `write_bytes` expects an implementation of `Write`. It would be nice to guarantee being error-free, but that would require code duplication.

File: rust-multiaddr/src/lib.rs (hunk 8 above)
Comment (tomaka) [reply rust-multiaddr/src/lib.rs]: Deprecated the function instead

File: rust-multiaddr/src/lib.rs (hunk 4 above)
Comment (Vurich) [reply rust-multiaddr/src/lib.rs]: @tomaka Can you either address this or explain why you don't want to address it?

File: rust-multiaddr/src/lib.rs (hunk 10 above)
Comment (Vurich) [reply rust-multiaddr/src/lib.rs]: @tomaka Same as above, either comment why this is fine or address it.

Comment (tomaka) [reply rust-multiaddr/src/lib.rs]: I made the change.

File: rust-multiaddr/src/lib.rs (hunk 4 above)
Comment (tomaka) [reply rust-multiaddr/src/lib.rs]: Done ; note that `ToString` is automatically implemented on `T where T: Display` by the stdlib.

**Review by eira-fransham — APPROVED** [review]
Looks great, glad we could get the last couple of things fixed.
